File: 152-maximum-product-subarray.c
/*
https://leetcode.com/problems/maximum-product-subarray/description/

152. Maximum Product Subarray

Find the contiguous subarray within an array (containing at least one number) which has the largest product.

For example, given the array [2,3,-2,4],
the contiguous subarray [2,3] has the largest product = 6.
*/

int maxProduct(int* nums, int numsSize) {
    int start = 0;
    int product = 1;
    int max = nums[0];
    for (int i = 0; i <= numsSize; i++) {
        if (i == numsSize || nums[i] == 0) {
            if (i < numsSize && nums[i] == 0 && max < 0) {
                max = 0;
            }
            
            for (int j = start; j < i-1; j++) {
                product = product / nums[j];
                if (product > max) {
                    max = product;
                }
                if (nums[j] < 0) {
                    break;
                }
            }

            start = i+1;
            product = 1;
            continue;
        }

        product = product * nums[i];
        if (product > max) {
            max = product;
        }
    }

    return max;
}

File: 152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        return self.solve_two_pointers(nums)
        
    def solve_two_pointers(self, nums):
        n = len(nums)
        if n == 0: return None

        product = 1
        maximum = max(nums)
        left, right = 0, 0

        while right <= n:
            # print(left, right, product, maximum)         
            if right == n or nums[right] == 0:
                while left < right-1:
                    product //= nums[left]
                    if maximum < product:
                        maximum = product
                    left += 1
                product = 1
                left = right + 1
                right += 1
                continue

            product *= nums[right]
            if maximum < product:
                maximum = product
            right += 1
        return maximum

def main():
    s = Solution()
    nums = [-2,0,-1]
    print(s.maxProduct(nums))

if __name__ == '__main__':
    main()